internal/pipeline/executor.go

```diff
@@ -0,0 +1,271 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pipeline
+
+import (
+	"fmt"
+	"path/filepath"
+
+	"go.uber.org/zap"
+	"sigs.k8s.io/kustomize/kyaml/kio"
+	"sigs.k8s.io/kustomize/kyaml/yaml"
+)
+
+// Executor hydrates a given pkg.
+type Executor struct {
+	PkgPath string
+	Logger  *zap.Logger
+}
+
+// Execute runs a pipeline.
+func (e *Executor) Execute() error {
+	logger := e.Logger.Sugar()
+
+	logger.Infof("Running pipeline for pkg %s", e.PkgPath)
+
+	path, err := filepath.Abs(e.PkgPath)
+	if err != nil {
+		return fmt.Errorf("failed to find abs path %s %w", e.PkgPath, err)
+	}
+
+	p := &pkg{
+		path: path,
+	}
+
+	// initialize hydration context
+	hctx := &hydrationContext{
+		root:            p,
+		logger:          logger,
+		hydrationStates: map[string]hydrationState{},
+		pkgs:            map[string]*pkg{},
+	}
+
+	resources, err := hydrate(p, hctx)
+	if err != nil {
+		return fmt.Errorf("failed to hydrate the pkg %s %w", path, err)
+	}
+	// TODO(droot): support different sink modes
+	// for now do in-place
+	logger.Infof("hydrated reources: %d", len(resources))
+
+	return nil
+}
+
+//
+// hydrationContext is a container to track state while hydrating a pkg.
+// This is sort of global state that is available to hydration process for
+// each pkg.
+type hydrationContext struct {
+	// root points to the root of hydration graph where we bagan our hydration journey
+	root *pkg
+
+	logger *zap.SugaredLogger
+
+	// TODO (droot): wire sink-mode here
+
+	// pkgs refers to the discovered pkgs where
+	// It's a map where key refers to package's absolute path and value
+	// refers to the pkg.
+	pkgs map[string]*pkg
+
+	// hydrationStates for each package (dry, hydrating or wet) in the hydration DAG
+	hydrationStates map[string]hydrationState
+}
+
+//
+// pkg represents a kpt package (node) in the hydration DAG.
+//
+// TODO(droot): kpt will certainly have pkg abstraction, may be embed that here to
+// avoid duplication.
+type pkg struct {
+	// absolute path on the filesystem
+	path string
+
+	// pipeline contained in the pkg
+	pipeline *Pipeline
+
+	// tracks the state of hydration (dry, hydrating, wet)
+	state hydrationState
+
+	// resources post hydration
+	resources []*yaml.RNode
+}
+
+// hydrationState represent hydration state of a pkg.
+type hydrationState int
+
+// constants for all the hydration states
+const (
+	Dry hydrationState = iota
+	Hydrating
+	Wet
+)
+
+func (s hydrationState) String() string {
```

Comment:
Consider using the golang stringer here: https://godoc.org/golang.org/x/tools/cmd/stringer

Reply:
hmmm...interesting. It does introduces dependency on `stringer` tool, wondering if it is worth the benefits (saving just one fn.)